### PR TITLE
Sfint 4124 - SubjectInput component

### PR DESCRIPTION
**Pull request by mmitiche**
## subjectInput component
subjectInput

### Demonstration of how the subjectInput will behave when the required property is set to true and the input is left  empty

https://user-images.githubusercontent.com/86681870/143309014-d24cd476-1448-44f5-afe4-6c8d9662af42.mov

### CODE REVIEW

File: src/main/default/lwc/subjectInput/subjectInput.html [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<template>
+  <label for="subject-input">
+    <c-section-title title={label}></c-section-title>
+  </label>
+  <div class="slds-is-relative">
+    <div class={formClass}>
```
Comment (nathanlb) [inline src/main/default/lwc/subjectInput/subjectInput.html:6]: Is it necessary to have this div and it's direct child div or could they be merged into one with `formClass` including "slds-form-element__control"?

File: src/main/default/lwc/subjectInput/subjectInput.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ * <c-subject-input label="Write a descriptive title" maxLength="50" message-when-value-missing="Complete this field." required></c-subject-input>
+ */
+export default class SubjectInput extends LightningElement {
+  labels = {
+    writeDescriptiveTitle,
+    errorValueMissing
+  };
+  /**
+   * The label of the input.
+   * @api
```
Comment (nathanlb) [inline src/main/default/lwc/subjectInput/subjectInput.js:17]: Just a reminder that these `@api` jsdoc tags might not be useful to you as it's a custom tag we added that's only used by quantic's custom jsdoc template. Feel free to keep them if you prefer though.

File: src/main/default/lwc/subjectInput/subjectInput.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      ? 'slds-form-element slds-has-error'
+      : 'slds-form-element';
+  }
+
+  /**
+   * Tells if there is an error in the input.
+   * @returns {boolean}
+   */
+  @api get hasError() {
+    return this._hasError;
```
Comment (nathanlb) [inline src/main/default/lwc/subjectInput/subjectInput.js:108]: could `this._hasError` be replaced with `!!this._errorMessage.length`? That way you can do away with the `_hasError` variable entirely.

**Review by nathanlb — APPROVED** [review]
Looks real nice! Just a few suggestions

File: src/main/default/lwc/subjectInput/subjectInput.js (hunk 2 above)
Comment (mmitiche) [reply src/main/default/lwc/subjectInput/subjectInput.js:17]: I will update this thanks👌

File: src/main/default/lwc/subjectInput/subjectInput.js (hunk 3 above)
Comment (mmitiche) [reply src/main/default/lwc/subjectInput/subjectInput.js:108]: Yes ! That's better👌

File: src/main/default/lwc/subjectInput/__test__/subjectInput.test.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  });
+
+  it('should display the input label', async () => {
+    const element = createTestComponent();
+    const expectedLabel = 'Expected Label';
+    element.label = expectedLabel;
+
+    await flushPromises();
+    const labelNode = element.shadowRoot.querySelector('c-section-title');
+    expect(labelNode).not.toBeNull();
```
Comment (jeremierobert-coveo) [inline src/main/default/lwc/subjectInput/__test__/subjectInput.test.js:49]: What do you think of ?
```suggestion
    expect(labelNode).toBeDefined();
```

File: src/main/default/lwc/subjectInput/__test__/subjectInput.test.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  });
+
+  it('should display the correct localized label', async () => {
+    const element = createTestComponent();
+    const expectedLabel = 'Write a descriptive title';
+
+    await flushPromises();
+    const labelNode = element.shadowRoot.querySelector('c-section-title');
+    expect(labelNode).not.toBeNull();
+    expect(labelNode.title).toBe(expectedLabel);
```
Comment (jeremierobert-coveo) [inline src/main/default/lwc/subjectInput/__test__/subjectInput.test.js:60]:
```suggestion
    const element = createTestComponent();

    await flushPromises();
    const labelNode = element.shadowRoot.querySelector('c-section-title');
    expect(labelNode).not.toBeNull();
    expect(labelNode.title).toBe('Write a descriptive title');
```

File: src/main/default/lwc/subjectInput/subjectInput.css [hunk 6]
```diff
@@ -0,0 +1,12 @@
+.input_count {
+  top: 0;
+  right: 0.75rem;
+  width: 3.5rem;
+  height: 2.5rem;
+  justify-content: flex-end;
+  font-weight: 200;
+}
+.slds-input {
```
Comment (jeremierobert-coveo) [inline src/main/default/lwc/subjectInput/subjectInput.css:9]:
```suggestion
}

.slds-input {
```

File: src/main/default/lwc/subjectInput/subjectInput.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+  /**
+   * Returns the CSS class of the form.
+   * @returns {string}
+   */
+  get formClass() {
+    return this._hasError
+      ? 'slds-form-element slds-has-error'
+      : 'slds-form-element';
+  }
```
Comment (jeremierobert-coveo) [inline src/main/default/lwc/subjectInput/subjectInput.js]:
```suggestion
  get formClass() {
    return `slds-form-element ${this._hasError ? 'slds-has-error' : ''}`;
  }
```

**Review by jeremierobert-coveo — APPROVED** [review]

File: src/main/default/lwc/subjectInput/subjectInput.html (hunk 1 above)
Comment (mmitiche) [reply src/main/default/lwc/subjectInput/subjectInput.html:6]: I followed the same structure from the SLDS blueprint but looking at what the CSS classes contains, in my case I can get rid of one of the divs so I will update this 👌

**Pull request merged**